metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1579,9 +1581,51 @@ class MetalsLspService(
     indexer.indexWorkspaceSources(buildTargets.allWritableData)
   }
 
-  def shutDownBloop(): Boolean = bloopServers.shutdownServer()
-
-  def isBloop(): Boolean = bspSession.exists(_.main.isBloop)
+  /**
+   * We only support restarting bloop and sbt servers.
+   */
+  def restartBspServer(): Future[Boolean] = {
+    def emitMessage(msg: String, msgType: MessageType = MessageType.Info) = {
+      languageClient.logMessage(new MessageParams(msgType, msg))
+    }
+    def handleRestartResult(result: Boolean, bsp: String) = {
+      if (!result) {
+        emitMessage(
+          s"Could not restart $bsp server.",
+          msgType = MessageType.Warning,
+        )
+      }
+      result
+    }
+    bspSession match {
+      case Some(session) if session.main.isBloop =>
+        val res = bloopServers.shutdownServer()
+        for {
+          _ <- autoConnectToBuildServer()
+        } yield handleRestartResult(res, "bloop")
+      case Some(session) if session.main.isSbt =>
+        for {
+          currentBuildTool <- supportedBuildTool
+          res <- currentBuildTool match {
+            case Some(sbt: SbtBuildTool) =>
+              for {
+                _ <- disconnectOldBuildServer()
+                code <- sbt.shutdownSbtServer(shellRunner, workspace = folder)
```

Comment:
Could we this for each server and return false if it was not possible?

Add `def shutdown() = false` method to BuildTool and override in SbtBuildTool and maybe MillBuildTool? `mill shutdown` should work there